src/main/cpp/utilities/diskwriter.cpp

```diff
@@ -23,35 +23,100 @@
 #include "diskwriter.h"
 #include "audioengine.h"
 #include "wavewriter.h"
+#include "wavereader.h"
 #include "utils.h"
+#include <stdio.h>
 #include <definitions/notifications.h>
 #include <messaging/notifier.h>
 
 namespace DiskWriter
 {
-    std::string outputDirectory;
-    unsigned long outputBufferSize  = 0;
-    unsigned long outputWriterIndex = 0;
-    AudioBuffer*  cachedBuffer      = nullptr;
-
-    /**
-     * prepare a new iteration of recording
-     */
-    void prepare( std::string aOutputDir, int aBufferSize, int amountOfChannels )
+    std::string              outputFile;
+    std::string              tempDirectory;
+    std::vector<writtenFile> outputFiles;
+    unsigned long            tempChunkSize     = 0;
+    unsigned long            outputWriterIndex = 0;
+    AudioBuffer*             cachedBuffer      = nullptr;
+
+    void prepare( std::string outputFilename, int chunkSize, int amountOfChannels )
     {
-        outputDirectory  = aOutputDir;
-        outputBufferSize = aBufferSize;
+        outputFile = outputFilename;
+
+        std::string path   = std::string( outputFilename );
+        std::size_t dirPos = path.find_last_of( "/" );
+        tempDirectory      = path.substr( 0, dirPos ) + "/";
 
+        tempChunkSize = chunkSize;
+
+        outputFiles.clear();
         generateOutputBuffer( amountOfChannels );
     }
 
+    bool finish()
+    {
+        if ( outputFiles.size() == 0 )
```

Comment:
Concatenates all snippets after recording is halted. This runs on the UI thread by default as it is called through JNI from MWEngine.java which is separate from the thread that runs the rendering. There is still an improvement to be had in calling _writeBufferToFile_ (which is called at the end of a render iteration and can stall on long writes).